RandBLAS/sparse.hh

```diff
@@ -123,110 +134,123 @@ SparseSkOp<T>::SparseSkOp(
+            vec_work[j] = swap;
             // update (rows, cols, vals)
-            sa_idxs[j + offset] = swap;
-            vals[j + offset] = (rout.v[1] % 2 == 0) ? 1.0 : -1.0;
-            la_idxs[j + offset] = i;
-
+            vec_ax_idxs[j + offset] = swap;
+            vals[j + offset] = (rv[1] % 2 == 0) ? 1.0 : -1.0;
+            rep_ax_idxs[j + offset] = i;
             // increment counter
-            impl_state.ctr.incr(1);
+            ctri.incr();
         }
-        // Restore sa_vec_work for next iteration of Fisher-Yates.
+        // Restore vec_work for next iteration of Fisher-Yates.
         //      This isn't necessary from a statistical perspective,
         //      but it makes it easier to generate submatrices of
         //      a given SparseSkOp.
-        for (j = 1; j <= k; ++j) {
-            int jj = k - j;
-            swap = sa_idxs[jj + offset];
-            ell = pivots[jj];
-            sa_vec_work[jj] = sa_vec_work[ell];
-            sa_vec_work[ell] = swap;
+        for (int64_t j = 1; j <= vec_nnz; ++j) {
+            int64_t jj = vec_nnz - j;
+            int64_t swap = vec_ax_idxs[jj + offset];
+            int64_t ell = pivots[jj];
+            vec_work[jj] = vec_work[ell];
+            vec_work[ell] = swap;
         }
-
-        out_state = impl_state;
+        ctr = ctri;
     }
-    S0.next_state = out_state;
-    return out_state;
+    return RNGState<RNG> {ctr, key};
 }
 
-
-template <typename T>
-RNGState fill_saso(
-    SparseSkOp<T> &S0
+template <typename SKOP>
+static
+auto fill_sparse(
+    SKOP & S0
 ) {
-    switch (S0.seed_state.rng_name) {
-        case RNGName::Philox:
-            return template_fill_saso<T, Philox>(S0);
-        case RNGName::Threefry:
-            return template_fill_saso<T, Threefry>(S0);
-        default:
-            throw std::runtime_error(std::string("Unrecognized generator."));
+    int64_t long_ax_len = MAX(S0.dist.n_rows, S0.dist.n_cols);
+    int64_t short_ax_len = MIN(S0.dist.n_rows, S0.dist.n_cols);
+
+    bool is_wide = S0.dist.n_rows == short_ax_len;
+    int64_t *short_ax_idxs = (is_wide) ? S0.rows : S0.cols;
+    int64_t *long_ax_idxs = (is_wide) ? S0.cols : S0.rows;
+
+    int64_t *vec_ax_idxs, *rep_ax_idxs;
+    int64_t vec_len, num_vecs;
+    if (S0.dist.family == SparseDistName::SASO) {
+        vec_len = short_ax_len;
+        num_vecs = long_ax_len;
+        vec_ax_idxs = short_ax_idxs;
+        rep_ax_idxs = long_ax_idxs;
+    } else {
+        vec_len = long_ax_len;
+        num_vecs = short_ax_len;
+        vec_ax_idxs = long_ax_idxs;
+        rep_ax_idxs = short_ax_idxs;
     }
+    S0.next_state = repeated_fisher_yates(
+        S0.seed_state, num_vecs, vec_len,
+        S0.dist.vec_nnz, vec_ax_idxs, rep_ax_idxs, S0.vals
+    );
+    return S0.next_state;
 }
 
-template <typename T>
-void print_saso(SparseSkOp<T>& S0) {
-    std::cout << "SASO information" << std::endl;
+template <typename SKOP>
+void print_saso(SKOP const& S0) {
```

Comment:
TODO: rename function to ``print_sparseskop`` or ``print_sparse_sk_op`` or similar.